src/tiled/objectrefdialog.cpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * ObjectRefdialog.cpp
+ * Copyright 2019, Thorbjørn Lindeijer <bjorn@lindeijer.nl>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "objectrefdialog.h"
+#include "ui_objectrefdialog.h"
+
+#include "documentmanager.h"
+#include "mapdocument.h"
+#include "map.h"
+#include "objectgroup.h"
+#include "mapobject.h"
+#include "utils.h"
+
+#include <QList>
+#include <QLineEdit>
+#include <QTableWidget>
+#include <QTableWidgetItem>
+#include <QTableWidgetSelectionRange>
+#include <QPushButton>
+#include <QHeaderView>
+
+namespace Tiled {
+
+
+ObjectRefDialog::ObjectRefDialog(QWidget *parent)
+    : QDialog(parent)
+    , mUi(new Ui::ObjectRefDialog)
+    , mId(0)
+{
+#if QT_VERSION < QT_VERSION_CHECK(5, 10, 0)
+    setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);
+#endif
+    mUi->setupUi(this);
+
+    Utils::restoreGeometry(this);
+
+    QTableWidget *tableWidget = mUi->tableWidget;
```

Comment:
See my comment at https://github.com/bjorn/tiled/pull/1408/files#r91926128

Reply:
As a first step I've replaced the `QTableWidget` with a `QTreeWidget`. Unfortunately, there appears to be a small problem with using `QTreeWidget`, in that it seems to automatically sort on the first column. In contrast the `QTableWidget` order remained stable until one of the headers was clicked. This would not be much of a problem, were it not for the fact that the
IDs are sorted lexicographically.

In the end, the plan remains to try replacing this with a `QTreeView` backed by `MapObjectModel`, proxied by a `ObjectsFilterModel` (currently in `objectsview.cpp`).